### DISCUSSION TITLE
More rustic expect_context

### DISCUSSION THREAD

**Discussion by geovie** [leptos-rs/leptos#3299] [Ideas]
With leptos 0.7 most of the api was moved from the `create_signal` to a more rust like `Signal::new` api.

One thing that still stands out a bit is `expect_context`.

Would it make sense to provide a trait + blanket impl similar to this:

```rust
trait ExpectContext {
    fn expect_context() -> Self;
}

impl<T: Clone + 'static> ExpectContext for T {
    fn expect_context() -> Self {
        expect_context()
    }
}
```

which would then allow to use it like the following:
```rust
// previously
    let theme = expect_context::<RwSignal::<Theme>>();
    let theme: RwSignal::<Theme> = expect_context();
// additionally with the above trait + impl
    let theme = RwSignal::<Theme>::expect_context();
```

Now I'm myself not convinced that it's better 😅 also because it looses the symmetry with `provide_context`, wdyt?

**Comment by gbj** [comment]
I'm not sure why that would be better or seems more idiomatic. Could you say more about why that seems to you to be the case?

**Reply by geovie** [reply]
I can't say for sure...
Maybe because it reminds me of the `Struct::new()` or `Struct::default()` syntax and it feels more natural to write `Struct::expect_context()` as well.
But it's not really constructing a new thing so that's probably not a good comparison.

Anyway I honestly don't have a strong opinion and just wanted to open the discussion to get some other opinions
